{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2b31ee45",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c84f8007",
   "metadata": {},
   "source": [
    "Brian thinks the protocol being used for Ali's groups long-RNA-seq experiments could be improved.\n",
    "\n",
    "Jennifer would like everything defined so they can just run the command if this change request comes in after the end of their ENCODE funding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "46aae881",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "from pathlib import Path\n",
    "import os\n",
    "import sys\n",
    "\n",
    "EC = str(Path(\"~/proj/encoded_client\").expanduser())\n",
    "if EC not in sys.path:\n",
    "    sys.path.append(EC)\n",
    "from encoded_client.encoded import ENCODED, HTTPError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6aeac0aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "known_document_types = {\n",
    "    '/documents/ecb1c489-905b-4afb-8389-ed7fb2e36ff2/': \"Chromium v2\",\n",
    "    '/documents/515a990c-50c4-459a-8036-502feaf5a18c/': \"Chromium v2 technical note\",\n",
    "    '/documents/3408b7a1-6b1b-45d1-a73e-db0556e6abdc/': \"Split-seq analysis protocol\",\n",
    "    '/documents/3cdf327b-df85-4149-9fca-a7541e6f1cb1/': \"Split-seq overview\",\n",
    "    '/documents/c155e58a-13ab-4af9-8591-cee13d639a40/': \"Nuclei Isolation protocol from tissue for 10x multiome\",\n",
    "    '/documents/53fb4382-c54f-4761-89bb-6a01a0dca97d/': \"Chromium GEM v3.1\",\n",
    "    '/documents/4d42b1ef-c9e6-4b31-a443-1266079545d0/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/d18bad0e-9a86-40f0-8ded-2c74072910e5/': \"Nuclei Isolation protocol from tissue for 10x multiome\",  # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/e3110414-4013-425e-94f6-72c549702bde/': \"Pool split protocol\",\n",
    "    '/documents/1bb75b62-ac29-4368-9855-68d410e1963a/': \"GM12878 Cell Culture protocol\",\n",
    "    '/documents/28fef8e0-e171-450d-b66d-e3d199531cf2/': \"Ambion spike in mixture\",\n",
    "    '/documents/a6e06058-5019-488f-9195-8bcd2e9229c1/': \"Laser capture microdisection\",\n",
    "    '/documents/c4b8952c-d9f1-42e7-8ea5-54659d14f46b/': \"Human tissue preservation protocol flash-freezing (version 1)\",\n",
    "    '/documents/c5c4174c-0f33-4793-be9d-f2b6a9357dae/': \"Nuclei Isolation protocol from tissue for 10x multiome\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/84a66218-cd6a-4959-adc0-91ecc513117f/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/c6235c97-a978-41b6-8a7e-1e16b983f4ba/': \"inDrop Sample Collection Protocol\",\n",
    "    '/documents/8179222b-9ad7-4274-a07e-820dbcac6aca/': \"inDrop Library Prep\",\n",
    "    '/documents/80953249-5d87-4f2c-84e0-0212c404c8d6/': \"Nuclei Isolation protocol from tissue 10x snRNA\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/57177fd0-da98-4e8f-95a4-459fbaeea146/': \"Chromium GEM v3.1\", # psame as 53fb4382-c54f-4761-89bb-6a01a0dca97d\n",
    "    '/documents/5f5193b0-0c61-40c6-a286-b1f6d6f0c0da/': \"SmartSeq\",\n",
    "    '/documents/9a7e281f-b901-422b-9f01-378ba4b9e97f/': \"Mouse adrenal gland nuclei isolation protocol for single nucleus RNA-seq\",\n",
    "    '/documents/f1b60b4e-b5d2-4b8d-af0b-06b767fa07c3/': \"Biosample purification protocol\",\n",
    "    '/documents/49fd5b3b-9878-4ec8-92bf-ef6dce9116e3/': \"SMARTer Ultra Low RNA kit\",\n",
    "    \"/documents/f16e44c9-ee9d-4ede-9bb5-886ea451d7a0/\": \"Cortex nuclei isolation protocol\",\n",
    "    \"/documents/eb624ae4-53ea-4de0-ba9e-11b40eb20e2c/\": \"cell isolation protocol\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1c8fdd41",
   "metadata": {},
   "outputs": [],
   "source": [
    "server = ENCODED(\"www.encodeproject.org\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6740e7dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "where = {\n",
    "    \"lab.title\": \"Ali Mortazavi, UCI\",\n",
    "    \"assay_title\": \"long read RNA-seq\",\n",
    "}\n",
    "query = server.search_jsonld(type=\"Experiment\", limit=\"500\", **where)\n",
    "\n",
    "def update_documents(seen_documents, experiment_id, obj):\n",
    "    for d in obj.get(\"documents\", []):\n",
    "        seen_documents.setdefault(d, []).append(experiment_id)\n",
    "\n",
    "seen_documents = {}\n",
    "for row in query[\"@graph\"]:\n",
    "    experiment_id = row[\"@id\"]\n",
    "    experiment = server.get_json(experiment_id)\n",
    "    update_documents(seen_documents, experiment_id, experiment)\n",
    "    for rep in experiment[\"replicates\"]:\n",
    "        library = rep[\"library\"]\n",
    "        update_documents(seen_documents, experiment_id, library)\n",
    "        if \"biosample\" in rep[\"library\"]:\n",
    "            biosample = rep[\"library\"][\"biosample\"]\n",
    "            update_documents(seen_documents, experiment_id, biosample)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3d52cc43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>used</th>\n",
       "      <th>filename</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://www.encodeproject.org/documents/6d583a...</td>\n",
       "      <td>39</td>\n",
       "      <td>ENCODE Long Read RNA-Seq Analysis Pipeline v3....</td>\n",
       "      <td>\"This document describes 1) the steps used to ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://www.encodeproject.org/documents/3baa46...</td>\n",
       "      <td>180</td>\n",
       "      <td>ENCODE_longread_wetlab_protocolv3.pdf</td>\n",
       "      <td>This protocol describes an optimized method fo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://www.encodeproject.org/documents/b556e2...</td>\n",
       "      <td>1</td>\n",
       "      <td>3348003_5X_DLPFC.tif</td>\n",
       "      <td>fresh frozen section, 30 microns, post fixed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://www.encodeproject.org/documents/d3cf9d...</td>\n",
       "      <td>1</td>\n",
       "      <td>7461192_5X_DLPFC.tif</td>\n",
       "      <td>fresh frozen section, 30 microns, post fixed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://www.encodeproject.org/documents/e90eb5...</td>\n",
       "      <td>1</td>\n",
       "      <td>4368365_5X_DLPFC.tif</td>\n",
       "      <td>fresh frozen section, 30 microns, post fixed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>https://www.encodeproject.org/documents/22a7ff...</td>\n",
       "      <td>1</td>\n",
       "      <td>6341028_5X_DLPFC.tif</td>\n",
       "      <td>fresh frozen section, 30 microns, post fixed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>https://www.encodeproject.org/documents/e7fcb8...</td>\n",
       "      <td>1</td>\n",
       "      <td>6207143_5X_DLPFC.tif</td>\n",
       "      <td>fresh frozen section, 30 microns, post fixed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>https://www.encodeproject.org/documents/303304...</td>\n",
       "      <td>1</td>\n",
       "      <td>7948794_5X_DLPFC.tif</td>\n",
       "      <td>fresh frozen section, 30 microns, post fixed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>https://www.encodeproject.org/documents/dcd089...</td>\n",
       "      <td>1</td>\n",
       "      <td>4163791_5X_DLPFC.tif</td>\n",
       "      <td>fresh frozen section, 30 microns, post fixed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>https://www.encodeproject.org/documents/d4da5f...</td>\n",
       "      <td>1</td>\n",
       "      <td>5194210_5X_DLPFC.tif</td>\n",
       "      <td>fresh frozen section, 30 microns, post fixed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>https://www.encodeproject.org/documents/ecd675...</td>\n",
       "      <td>1</td>\n",
       "      <td>5115600_5X_DLPFC.tif</td>\n",
       "      <td>fresh frozen section, 30 microns, post fixed i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>https://www.encodeproject.org/documents/ae021f...</td>\n",
       "      <td>14</td>\n",
       "      <td>Macrophage_differentiation_and_M1_M2_activatio...</td>\n",
       "      <td>HL-60 M0/M1/M2 differentiation protocol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>https://www.encodeproject.org/documents/b88f13...</td>\n",
       "      <td>7</td>\n",
       "      <td>Direct-RNA_Wet-Lab_DCC v1.0.pdf</td>\n",
       "      <td>ONT dRNA experimental protocol v1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>https://www.encodeproject.org/documents/cc4b42...</td>\n",
       "      <td>7</td>\n",
       "      <td>Direct-RNA_Computational-Analysis v1.0.pdf</td>\n",
       "      <td>ONT dRNA computational protocol v1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>https://www.encodeproject.org/documents/81af56...</td>\n",
       "      <td>63</td>\n",
       "      <td>ENCODE Long Read RNA-Seq Analysis Pipeline v3 ...</td>\n",
       "      <td>This document describes 1) the steps used to g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>https://www.encodeproject.org/documents/e90954...</td>\n",
       "      <td>15</td>\n",
       "      <td>ENCODE_Protocol_Spikeins_capping_v1.pdf</td>\n",
       "      <td>Protocol to add 5’ cap structures to exogenous...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>https://www.encodeproject.org/documents/bf543f...</td>\n",
       "      <td>38</td>\n",
       "      <td>ENCODE Long Read RNA-Seq Analysis Pipeline v3....</td>\n",
       "      <td>Computational protocol for processing mouse lo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>https://www.encodeproject.org/documents/7ca4f1...</td>\n",
       "      <td>54</td>\n",
       "      <td>ENCODE Long Read RNA-Seq Analysis Pipeline v3....</td>\n",
       "      <td>Computational protocol for processing mouse lo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>https://www.encodeproject.org/documents/c4b895...</td>\n",
       "      <td>33</td>\n",
       "      <td>ENCODE4_Snyder_HumanTissuePreservation_Flash_v...</td>\n",
       "      <td>Human tissue preservation protocol flash-freez...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>https://www.encodeproject.org/documents/77db75...</td>\n",
       "      <td>55</td>\n",
       "      <td>ENCODE_protocol_2020pacbio_Final.pdf</td>\n",
       "      <td>This protocol describes an optimized method fo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>https://www.encodeproject.org/documents/0a54a6...</td>\n",
       "      <td>38</td>\n",
       "      <td>ENCODE Long Read RNA-Seq Analysis Pipeline v3....</td>\n",
       "      <td>PacBio RNA-seq computational analysis pipeline...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>https://www.encodeproject.org/documents/5a9cb0...</td>\n",
       "      <td>10</td>\n",
       "      <td>ENCODE Long Read RNA-Seq Analysis Pipeline (Mo...</td>\n",
       "      <td>The mouse long-read RNA-seq data on the ENCODE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>https://www.encodeproject.org/documents/54012f...</td>\n",
       "      <td>8</td>\n",
       "      <td>Protocol_ENCODEsubmission_nosizeselected-2.pdf</td>\n",
       "      <td>Protocol for creating non-size selected cDNA l...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>https://www.encodeproject.org/documents/229723...</td>\n",
       "      <td>7</td>\n",
       "      <td>ENCODE_protocol_brain_illumina.pdf</td>\n",
       "      <td>Protocol for creating cDNA libraries from brai...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>https://www.encodeproject.org/documents/7ec9d6...</td>\n",
       "      <td>9</td>\n",
       "      <td>ENCODE Long Read RNA-Seq Analysis Pipeline (Hu...</td>\n",
       "      <td>The human long-read RNA-seq data on the ENCODE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>https://www.encodeproject.org/documents/9d64c0...</td>\n",
       "      <td>2</td>\n",
       "      <td>Protocol_ENCODEsubmission_nosizeselected (1).pdf</td>\n",
       "      <td>Protocol for generating PacBio Iso-seq librari...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>https://www.encodeproject.org/documents/fc272a...</td>\n",
       "      <td>1</td>\n",
       "      <td>ENCODE PacBio Iso-seq Analysis Pipeline_ GM128...</td>\n",
       "      <td>Protocol describing computational processing o...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   id  used  \\\n",
       "0   https://www.encodeproject.org/documents/6d583a...    39   \n",
       "1   https://www.encodeproject.org/documents/3baa46...   180   \n",
       "2   https://www.encodeproject.org/documents/b556e2...     1   \n",
       "3   https://www.encodeproject.org/documents/d3cf9d...     1   \n",
       "4   https://www.encodeproject.org/documents/e90eb5...     1   \n",
       "5   https://www.encodeproject.org/documents/22a7ff...     1   \n",
       "6   https://www.encodeproject.org/documents/e7fcb8...     1   \n",
       "7   https://www.encodeproject.org/documents/303304...     1   \n",
       "8   https://www.encodeproject.org/documents/dcd089...     1   \n",
       "9   https://www.encodeproject.org/documents/d4da5f...     1   \n",
       "10  https://www.encodeproject.org/documents/ecd675...     1   \n",
       "11  https://www.encodeproject.org/documents/ae021f...    14   \n",
       "12  https://www.encodeproject.org/documents/b88f13...     7   \n",
       "13  https://www.encodeproject.org/documents/cc4b42...     7   \n",
       "14  https://www.encodeproject.org/documents/81af56...    63   \n",
       "15  https://www.encodeproject.org/documents/e90954...    15   \n",
       "16  https://www.encodeproject.org/documents/bf543f...    38   \n",
       "17  https://www.encodeproject.org/documents/7ca4f1...    54   \n",
       "18  https://www.encodeproject.org/documents/c4b895...    33   \n",
       "19  https://www.encodeproject.org/documents/77db75...    55   \n",
       "20  https://www.encodeproject.org/documents/0a54a6...    38   \n",
       "21  https://www.encodeproject.org/documents/5a9cb0...    10   \n",
       "22  https://www.encodeproject.org/documents/54012f...     8   \n",
       "23  https://www.encodeproject.org/documents/229723...     7   \n",
       "24  https://www.encodeproject.org/documents/7ec9d6...     9   \n",
       "25  https://www.encodeproject.org/documents/9d64c0...     2   \n",
       "26  https://www.encodeproject.org/documents/fc272a...     1   \n",
       "\n",
       "                                             filename  \\\n",
       "0   ENCODE Long Read RNA-Seq Analysis Pipeline v3....   \n",
       "1               ENCODE_longread_wetlab_protocolv3.pdf   \n",
       "2                                3348003_5X_DLPFC.tif   \n",
       "3                                7461192_5X_DLPFC.tif   \n",
       "4                                4368365_5X_DLPFC.tif   \n",
       "5                                6341028_5X_DLPFC.tif   \n",
       "6                                6207143_5X_DLPFC.tif   \n",
       "7                                7948794_5X_DLPFC.tif   \n",
       "8                                4163791_5X_DLPFC.tif   \n",
       "9                                5194210_5X_DLPFC.tif   \n",
       "10                               5115600_5X_DLPFC.tif   \n",
       "11  Macrophage_differentiation_and_M1_M2_activatio...   \n",
       "12                    Direct-RNA_Wet-Lab_DCC v1.0.pdf   \n",
       "13         Direct-RNA_Computational-Analysis v1.0.pdf   \n",
       "14  ENCODE Long Read RNA-Seq Analysis Pipeline v3 ...   \n",
       "15            ENCODE_Protocol_Spikeins_capping_v1.pdf   \n",
       "16  ENCODE Long Read RNA-Seq Analysis Pipeline v3....   \n",
       "17  ENCODE Long Read RNA-Seq Analysis Pipeline v3....   \n",
       "18  ENCODE4_Snyder_HumanTissuePreservation_Flash_v...   \n",
       "19               ENCODE_protocol_2020pacbio_Final.pdf   \n",
       "20  ENCODE Long Read RNA-Seq Analysis Pipeline v3....   \n",
       "21  ENCODE Long Read RNA-Seq Analysis Pipeline (Mo...   \n",
       "22     Protocol_ENCODEsubmission_nosizeselected-2.pdf   \n",
       "23                 ENCODE_protocol_brain_illumina.pdf   \n",
       "24  ENCODE Long Read RNA-Seq Analysis Pipeline (Hu...   \n",
       "25   Protocol_ENCODEsubmission_nosizeselected (1).pdf   \n",
       "26  ENCODE PacBio Iso-seq Analysis Pipeline_ GM128...   \n",
       "\n",
       "                                          description  \n",
       "0   \"This document describes 1) the steps used to ...  \n",
       "1   This protocol describes an optimized method fo...  \n",
       "2   fresh frozen section, 30 microns, post fixed i...  \n",
       "3   fresh frozen section, 30 microns, post fixed i...  \n",
       "4   fresh frozen section, 30 microns, post fixed i...  \n",
       "5   fresh frozen section, 30 microns, post fixed i...  \n",
       "6   fresh frozen section, 30 microns, post fixed i...  \n",
       "7   fresh frozen section, 30 microns, post fixed i...  \n",
       "8   fresh frozen section, 30 microns, post fixed i...  \n",
       "9   fresh frozen section, 30 microns, post fixed i...  \n",
       "10  fresh frozen section, 30 microns, post fixed i...  \n",
       "11            HL-60 M0/M1/M2 differentiation protocol  \n",
       "12                ONT dRNA experimental protocol v1.0  \n",
       "13               ONT dRNA computational protocol v1.0  \n",
       "14  This document describes 1) the steps used to g...  \n",
       "15  Protocol to add 5’ cap structures to exogenous...  \n",
       "16  Computational protocol for processing mouse lo...  \n",
       "17  Computational protocol for processing mouse lo...  \n",
       "18  Human tissue preservation protocol flash-freez...  \n",
       "19  This protocol describes an optimized method fo...  \n",
       "20  PacBio RNA-seq computational analysis pipeline...  \n",
       "21  The mouse long-read RNA-seq data on the ENCODE...  \n",
       "22  Protocol for creating non-size selected cDNA l...  \n",
       "23  Protocol for creating cDNA libraries from brai...  \n",
       "24  The human long-read RNA-seq data on the ENCODE...  \n",
       "25  Protocol for generating PacBio Iso-seq librari...  \n",
       "26  Protocol describing computational processing o...  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records = []\n",
    "for d in seen_documents:\n",
    "    document = server.get_json(d)\n",
    "    records.append({\n",
    "        \"id\": \"https://www.encodeproject.org{}\".format(d),\n",
    "        \"used\": len(seen_documents[d]),\n",
    "        \"filename\": document[\"attachment\"][\"download\"],\n",
    "        \"description\": document[\"description\"],\n",
    "    })\n",
    "    \n",
    "document_table = pandas.DataFrame(records)\n",
    "document_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a2143591",
   "metadata": {},
   "outputs": [],
   "source": [
    "document_table.to_csv(\"long-read-documents.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bfd11a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "seen_documents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "91e811d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " biosample_nic.csv         purkinje-cross-reference.csv\r\n",
      " long-read-documents.csv  'Snyder tissue correspondence.csv'\r\n",
      " merge_stam_wold.csv       who_made_what.csv\r\n"
     ]
    }
   ],
   "source": [
    "ls *.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ace9178a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "amarana\r\n"
     ]
    }
   ],
   "source": [
    "!hostname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b1f83f0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
